### PR TITLE
Use relative paths for archives

### PR DESCRIPTION
**Pull request by raymondbutcher**
## what
* Avoid using absolute paths which are not portable across workstations

## why

This fixes issues when multiple people are working on the same project and they have the project located in different places.

Before this change, we were seeing plans like:

```
  ~ module.lambda_ami_backup.aws_lambda_function.ami_backup
      filename:                    "/home/user1/some-project/terraform/workspace/management-nonprod/.terraform/modules/49890e3df5623d30b47ac70e1936d38f/ami_backup.zip" => "/Users/user2/some-project/terraform/workspace/management-nonprod/.terraform/modules/49890e3df5623d30b47ac70e1936d38f/ami_backup.zip"
```

We've tested this change and the plans are now clean.

I've used relative paths like this in other modules/projects, putting archives in the `.terraform` directory, and haven't had any issues so far. An md5 is used in the filename to avoid conflicts with multiple versions of this module being used in the same project (unlikely but it's better to be safe).

### CODE REVIEW

**Comment by osterman** [comment]
@raymondbutcher thanks for this fix! Makes sense. This was a bug =)

File: main.tf [hunk 1]
```diff
@@ -47,13 +47,13 @@ data "aws_iam_policy_document" "ami_backup" {
 data "archive_file" "ami_backup" {
   type        = "zip"
   source_file = "${path.module}/ami_backup.py"
-  output_path = "${path.module}/ami_backup.zip"
+  output_path = ".terraform/terraform-aws-ec2-ami-backup/backup-${md5(file("${path.module}/ami_backup.py"))}.zip"
```
Comment (osterman) [inline main.tf]: I'm concerned with the portability of hardcoding both `.terraform` and the module name `terraform-aws-ec2-ami-backup` which is subject to change in forks. 

I propose the following:
- [ ] create a new variable called `tfstate_dir` and default it to `.terraform`
- [ ] use a `local` called `module_name` which is derived from `basename(path.module)`
- [ ] use a `local` to define an `output_dir` which is equal to `${tfstate_dir}/${module_name}`

**Review by osterman — CHANGES_REQUESTED** [review]
See inline comments.

File: main.tf (hunk 1 above)
Comment (osterman) [reply main.tf]: This is also a nice fix: https://github.com/hashicorp/terraform/issues/8204#issuecomment-332239294

Take your pick =)

**Comment by raymondbutcher** [comment]
Thanks @osterman, fair points.

I've done a (nicer IMO) variation of https://github.com/hashicorp/terraform/issues/8204#issuecomment-332239294 that was suggested and this is the resulting plan:

```
  ~ module.lambda_ami_backup.aws_lambda_function.ami_backup
      filename: ".terraform/terraform-aws-ec2-ami-backup/backup-66f40c060f29561cb074987a03c45117.zip" => ".terraform/modules/49890e3df5623d30b47ac70e1936d38f/ami_backup.zip"

  ~ module.lambda_ami_backup.aws_lambda_function.ami_cleanup
      filename: ".terraform/terraform-aws-ec2-ami-backup/cleanup-ff78a3eaea904d4807a069837535acc0.zip" => ".terraform/modules/49890e3df5623d30b47ac70e1936d38f/ami_cleanup.zip"
```

As per my commit message:

> This removes the assumption about Terraform making a .terraform directory in the current working directory, and removes the need to use md5 to handle multiple module versions in the same project as Terraform deals with that itself. It now assumes that the module is located within the current working directory.

I think it's safe to assume that. https://github.com/hashicorp/terraform/issues/8204#issuecomment-332239294 makes the same assumption.

**Review by osterman — APPROVED** [review]

**Comment by osterman** [comment]
We have one other PR we should merge before this one. We will take care of this today.

**Pull request merged**